include/libtwofloat/arithmetics/double-word-arithmetic.hpp

```diff
@@ -233,5 +233,628 @@ inline two<T> div(const two<T> &x, const two<T> &y) {
+    p1 = qd::two_sqr(input, p2);
+    return two<T>{p1, p2};
+  }
+
+  // Reference: QD / dd_inline.h
+  // TODO: make sure this has not been already implemented
+  template <typename T>
+  inline two<T> add(T a, T b) {
+    T s, e;
+    s = qd::two_sum(a, b, e);
+    return two<T> (s, e);
+  }
+
+} // End namespace dd_real
+
+// Reference: QD / dd_inline.h
+/* Round to nearest integer */
+template <typename T>
+inline two<T> nint(const two<T> &input) {
+  T hi = qd::nint(input.h);
+  T lo;
+
+  T zeropointzero, pointfive, onepointzero;
+  if constexpr (std::is_same_v<T, float>) {
+    zeropointzero = 0.0f;
+    pointfive = 0.5f;
+    onepointzero = 1.0f;
+  } else if constexpr (std::is_same_v<T, double>) {
+    zeropointzero = 0.0;
+    pointfive = 0.5;
+    onepointzero = 1.0;
+  } else {
+    std::cerr << "Other types not supported\n";
+  }
+
+  if(hi == input.h) {
+    /* High word is an integer already. Round the low word. */
+    lo = qd::nint(input.l);
+
+    /* Renormalize. This is needed if h = some integer, l = 1/2. */
+    hi = qd::quick_two_sum(hi, lo, lo);
+  }
+  else {
+    /* High word is not an integer */
+    lo = zeropointzero;
+    if(std::abs(hi - input.h) == pointfive && input.l < zeropointzero) {
+      /* There is a tie in the high word, consult the low word
+         to break the tie. */
+      hi -= onepointzero; /* NOTE: This does not cause INEXACT. */
+    }
+  }
+
+  return two<T>{hi, lo};
+}
+
+// Reference: QD / dd_inline.h
+/* Cast to double. */
+template <typename T>
+inline T to_double(const two<T> &input) {
+  return input.h;
+}
+
+// Reference: QD / dd_inline.h
+template <typename T>
+inline two<T> sqr(const two<T> &input) {
+
+  T twopointzero;
+  if constexpr (std::is_same_v<T, float>) {
+    twopointzero = 2.0f;
+  } else if constexpr (std::is_same_v<T, double>) {
+    twopointzero = 2.0;
+  } else {
+    std::cerr << "Other types not supported\n";
+  }
+
+  T p1, p2;
+  T s1, s2;
+  p1 = qd::two_sqr(input.h, p2);
+  p2 += twopointzero * input.h * input.l;
+  p2 += input.l * input.l;
+  s1 = qd::quick_two_sum(p1, p2, s2);
+  return two<T>{s1, s2};
+}
+
+// Reference: QD / dd_real.cpp
+/* Computes sin(a) using Taylor series.
+   Assumes |a| <= pi/32 */
+template<typename T>
+static two<T> sin_taylor(const two<T> &input) {
+
+  T pointfive, local_eps;
+  const T* local_ptr_inv_fact;
+  T zeropointzero;
+  if constexpr (std::is_same_v<T, float>) {
+    pointfive = 0.5f;
+    local_eps = fp32_eps;
+    local_ptr_inv_fact = &fp32_inv_fact[0][0];
+    zeropointzero = 0.0f;
+  } else if constexpr (std::is_same_v<T, double>) {
+    pointfive = 0.5;
+    local_eps = fp64_eps;
+    local_ptr_inv_fact = &fp64_inv_fact[0][0];
+    zeropointzero = 0.0;
+  } else {
+    std::cerr << "Other types not supported\n";
+  }
+
+  const T thresh = pointfive * std::abs(to_double(input)) * local_eps;
+
+  two<T> r, s, t, x;
+
+  if (input.eval() == zeropointzero) {
+    T hi, lo;
+    qd::split(zeropointzero, hi, lo);
```

Comment:
This is an expensive way to set T.h and T.l to 0!

Reply:
Removed this and other compute-intensive statements in commit [28d95dd](https://github.com/esa-tu-darmstadt/twofloat/pull/1/commits/28d95dd609f0dd902331a47db5b276b7b50c0f51).